subsys/fs/zms/zms.c

```diff
@@ -0,0 +1,1634 @@
+	 * Restore the ate_wra of the current sector after this
+	 */
+	previous_ate_wra = fs->ate_wra;
+#ifdef CONFIG_ZMS_LOOKUP_CACHE
+	zms_lookup_cache_invalidate(fs, sec_addr >> ADDR_SECT_SHIFT);
+#endif
+	rc = zms_add_empty_ate(fs, sec_addr);
+	fs->ate_wra = previous_ate_wra;
+
+	return rc;
+}
+
+int zms_clear(struct zms_fs *fs)
+{
+	int rc;
+	uint64_t addr;
+	bool ebw_required =
+		flash_params_get_erase_cap(fs->flash_parameters) & FLASH_ERASE_C_EXPLICIT;
+
+	if (!fs->ready) {
+		LOG_ERR("zms not initialized");
+		return -EACCES;
+	}
+
+	for (uint32_t i = 0; i < fs->sector_count; i++) {
+		addr = (uint64_t)i << ADDR_SECT_SHIFT;
+		if (ebw_required) {
+			rc = zms_flash_erase_sector(fs, addr);
+			if (rc) {
+				return rc;
+			}
+		}
+		rc = zms_add_empty_ate(fs, addr);
+		if (rc) {
+			return rc;
+		}
+	}
+
+	/* zms needs to be reinitialized after clearing */
+	fs->ready = false;
+
+	return 0;
+}
+
+static int zms_init(struct zms_fs *fs)
+{
+	int rc;
+	struct zms_ate last_ate, first_ate, close_ate, empty_ate;
+	size_t ate_size;
+	/* Initialize addr to 0 for the case fs->sector_count == 0. This
+	 * should never happen as this is verified in zms_mount() but both
+	 * Coverity and GCC believe the contrary.
+	 */
+	uint64_t addr = 0U;
+	uint32_t i, closed_sectors = 0;
+	bool ebw_required =
+		flash_params_get_erase_cap(fs->flash_parameters) & FLASH_ERASE_C_EXPLICIT;
+
+	k_mutex_lock(&fs->zms_lock, K_FOREVER);
+
+	ate_size = zms_al_size(fs, sizeof(struct zms_ate));
+	/* step through the sectors to find a open sector following
+	 * a closed sector, this is where zms can write.
+	 */
+
+	for (i = 0; i < fs->sector_count; i++) {
+		addr = ((uint64_t)i << ADDR_SECT_SHIFT) + (uint32_t)(fs->sector_size - ate_size);
+
+		/* read the first and second ate to find out if sector is
+		 * open/close
+		 */
+		rc = zms_get_sector_header(fs, addr, &empty_ate, &close_ate);
+		if (rc) {
+			goto end;
+		}
+
+		/* update cycle count */
+		fs->sector_cycle = empty_ate.cycle_cnt;
+
+		if (zms_close_ate_valid(fs, &close_ate) && zms_empty_ate_valid(fs, &empty_ate) &&
+		    (close_ate.cycle_cnt == empty_ate.cycle_cnt)) {
+			/* closed sector */
+			closed_sectors++;
+			/* Let's check first that we support this ZMS version */
+			if (ZMS_GET_VERSION(empty_ate.metadata) != ZMS_DEFAULT_VERSION) {
+				LOG_ERR("ZMS Version is not supported");
+				rc = -ENOEXEC;
+				goto end;
+			}
+			zms_sector_advance(fs, &addr);
+			/* addr is pointing to the close ATE */
+			rc = zms_get_sector_header(fs, addr, &empty_ate, &close_ate);
+			if (rc) {
+				goto end;
+			}
+			/* update cycle count */
+			fs->sector_cycle = empty_ate.cycle_cnt;
+			if (!zms_empty_ate_valid(fs, &empty_ate) ||
+			    !zms_close_ate_valid(fs, &close_ate) ||
+			    (close_ate.cycle_cnt != empty_ate.cycle_cnt)) {
+				/* open sector */
+				break;
+			}
+		}
+	}
+	/* all sectors are closed, this is not a zms fs */
+	if (closed_sectors == fs->sector_count) {
```

Comment:
Now if you follow with the ZMS identification  above, when you get to point like this you know that you are probably dealing with ZMS but something is not ok. This means that you may still be able to recover some of data, because this is probably ZMS.